contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,288 @@
+https://github.com/kubernetes/kubernetes/issues/40476.
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual
+objects, by sending individual GET requests. However, it is sending those requests
+periodically to enable automatic updates of mounted secrets/configmaps/...
+In large clusters, this is generating huge unnecessary load, as this load in
+principle should be watch-based. We would like to address this together with
+solving the authorization issue.
+
+
+# Proposal
+
+ In this proposal, we're not focusing on how exactly security should be done.
+We're just sketching very high level approach and exact authorization mechanism
+should be discussed separately.
+
+ At the high level, what we would like to achieve is to enable LIST and WATCH
+requests to support more sophisticated filtering (e.g. we would like to be able
+to somehow ask for all Secrets attached to all pods bound to a given node in a
+single LIST or WATCH request). We obviously want to be able to authorize all
+other types of requests (in particular GETs), so that the design has to be
+consistent in all those cases.
+
+ To solve this problem, we propose to introduce the idea of ```bulk watch ``` and
+```bulk get ``` (and in the future other bulk operations). This idea was already
+appearing in the community and now it seems to be a good time to proceed with it.
+
+ We would like the API to be generic enough so that it will be useful for many
+different usecases, not just for this particular one. As a result, below we are
+describing requirements that the solution has to meet to satisfy our needs:
+- a single bulk request has to support multiple resources (e.g. get a node and
+all pods associated with it)
+- if the bulk request filters asks only for object of a single resource type,
+the returned result should ideally be comparible with what is returned from
+the regular API
+
+TODO: What else do we require?
+
+
+# Detailed design
+
+ To make the API generic enough, one of our requirements is to make those bulk
+operations work across different resource types (e.g. watch pod A and secret S
+within a single watch call). [Though, this doesn't need to be supported by the
+initial implementation - this just have to be possible at the API level.]
+
+ In the current API, it is not possible to express a request that is spanning
+multiple different resource types. So as a result, we need to introduce completely
+new API to support it.
+
+ We will solve it by introducing a new dedicated API group:
+```
+/apis/bulk.k8s.io/
+```
+ that underneath will have a separate implementation. Initially we are going
+to support the following two operations:
+```
+POST /apis/bulk.k8s.io/get   <body defines filtering>
```

Comment:
These would also be versioned, so `/apis/bulk.k8s.io/v1/get`.  In a sense, bulk retrieval is an "aggregation API".  The default verb is POST, the resource is `get` or `bulkget` or `multiget`, and the serialization for `get` can be `BulkGet` / `BulkWatch`.  I don't actually know that we need two resources - we may want to try to model watch as `POST /apis/bulk.k8s.io/get?watch=1` or as a field in the body.  The response to a POST can be anything that GET supports, so we don't need two separate actions if LIST and WATCH need to support the same args.